pallets/randomness/src/lib.rs

```diff
@@ -0,0 +1,324 @@
+	use frame_support::pallet_prelude::*;
+	use frame_support::traits::{Currency, ReservableCurrency};
+	use frame_support::weights::WeightToFeePolynomial;
+	use frame_system::pallet_prelude::*;
+	use sp_runtime::traits::Saturating;
+	use sp_std::vec::Vec;
+
+	/// Request identifier, unique per request for randomness
+	pub type RequestId = u64;
+
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// Overarching event type
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+		/// Currency in which the security deposit will be taken.
+		type Currency: Currency<Self::AccountId> + ReservableCurrency<Self::AccountId>;
+		/// Get relay chain epoch index to insert into this pallet
+		type RelayEpochIndex: GetEpochIndex<u64>;
+		/// Get relay chain randomness to insert into this pallet
+		type RelayRandomness: GetRelayRandomness<Self::Hash>;
+		/// Send randomness to smart contract
+		/// TODO: why can't Randomness = T::Hash?
+		type RandomnessSender: SendRandomness<Self::AccountId, [u8; 32]>;
+		/// Convert a weight value into a deductible fee based on the currency type.
+		type WeightToFee: WeightToFeePolynomial<Balance = BalanceOf<Self>>;
+		#[pallet::constant]
+		/// The amount that should be taken as a security deposit when requesting randomness.
+		type Deposit: Get<BalanceOf<Self>>;
+		#[pallet::constant]
+		/// Requests expire and can be purged from storage after this many blocks
+		type ExpirationDelay: Get<Self::BlockNumber>;
+		// /// Weight information for extrinsics in this pallet.
+		// type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		NotYetImplemented,
+		RequestCounterOverflowed,
+		NotSufficientDeposit,
+		CannotRequestPastRandomness,
+		CannotBeFulfilledBeforeExpiry,
+		RequestDNE,
+		RequestCannotYetBeFulfilled,
+		OnlyRequesterCanIncreaseFee,
+		NewFeeMustBeGreaterThanOldFee,
+		RequestHasNotExpired,
+		RequestedRandomnessNotCorrectlyUpdated,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Randomness requested and request put in storage
+		RandomnessRequested {
+			id: RequestId,
+			refund_address: T::AccountId,
+			contract_address: T::AccountId,
+			fee: BalanceOf<T>,
+			salt: T::Hash,
+			info: RequestType<T::BlockNumber>,
+		},
+		RequestFulfilled {
+			id: RequestId,
+		},
+		RequestFeeIncreased {
+			id: RequestId,
+			new_fee: BalanceOf<T>,
+		},
+		RequestExpirationExecuted {
+			id: RequestId,
+		},
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn request)]
+	/// Randomness requests not yet fulfilled or purged
+	pub type Requests<T: Config> =
+		StorageMap<_, Blake2_128Concat, RequestId, RequestState<T>, OptionQuery>;
```

Comment:
`RequestId` is auto-incrementing and isn't otherwise user-influenceable, so I think we could use twox here